core/classifiers/classifier_utils.py

```diff
@@ -60,15 +61,152 @@ def extract_svm_parameters(clf):
+                new_items.append(item)
+            elif isinstance(item, (list)):
+                for subitem in item:
+                    new_items.append(subitem)
+            else:
+                raise Exception(
+                    'Exepcted all the values in list to be either stirngs, '
+                    'floats, integers, lists or dicts but '
+                    'received %s.' % type(item))
+        current_items = new_items
+        if not current_items:
+            break
+        item = current_items[0]
+        item_type = type(item)
+        if not all(isinstance(subitem, item_type) for subitem in current_items):
+            raise Exception(
+                'Expected all the values in list to have same type.')
+        if any(isinstance(item, t) for t in leaf_node_item_types):
+            break
+
+    if item_type == dict:
+        keys_value_type_dict = {
+            k: type(v) for k, v in current_items[0].iteritems()
+        }
+        for item in current_items:
+            item_value_type_dict = {k: type(v) for k, v in item.iteritems()}
+            if keys_value_type_dict != item_value_type_dict:
+                raise Exception(
+                    'Expected all the subdicts to have same set of keys '
+                    'and corresponding value types.')
+    return item_type
+
+
+def convert_float_numbers_to_string_in_classifier_data(classifier_data):
+    """Converts all floating point numbers in classifier data to string.
+
+    Args:
+        classifier_data: dict|list. The original classifier data which needs
+            conversion of floats to strings.
+
+    Returns:
+        dict|list. Modified classifier data in which float values are converted
+            into strings and each dict/subdict is augmented with a special key
+            'float_values' which contains list of keys whose values have
+            undergone the transformation.
+    """
+    # pylint: disable=too-many-branches
+    if isinstance(classifier_data, dict):
+        if vmconf.FLOAT_INDICATOR_KEY in classifier_data:
+            raise Exception(
+                'Classifier data already contains a %s key' %
+                vmconf.FLOAT_INDICATOR_KEY)
+        float_fields = []
+        for k in classifier_data:
+            if isinstance(classifier_data[k], (basestring, int)):
+                classifier_data[k] = classifier_data[k]
+            elif isinstance(classifier_data[k], dict):
+                classifier_data[k] = (
+                    convert_float_numbers_to_string_in_classifier_data(
+                        classifier_data[k]))
+            elif isinstance(classifier_data[k], list):
+                leaf_node_item_type = make_sure_that_list_has_uniform_structure(
+                    classifier_data[k])
+                new_list = (
+                    convert_float_numbers_to_string_in_classifier_data(
+                        classifier_data[k]))
+                if leaf_node_item_type in (float, dict):
```

Comment:
Not sure about appending to float_fields when it's a dict?

Reply:
Because, each subdict is having a `float_fields` key. So appending the key is required so that decoder can go through each subdict and transform any required value.

Reply:
But why wouldn't the decoder just recurse fully anyway, once it detects a non-primitive structure?

Reply:
The decoder is driven by `float_fields` and also it won't be able to detect dicts if they are part of a sublist of the list. So decoder would only see a list which is part of another list but not the dict. That's why we need to add that key to floats. So decoder does traverse that path. If all leaf are strings then there is no need to traverse and so key must be added when we want decoder to traverse.